Issue templates (#1425)
GPflow/GPflow: .github/ISSUE_TEMPLATE/bugs-performance-issues.md

---
name: Bug report (including performance and build issues)

> about: Help us improve the GPflow code base

Reviewer: > Help us improve the GPflow code base
This is true of most ticket types, right. Can we say something more specific, like
```suggestion
about: Help us find mistakes in the GPflow code base
```

> labels: bug

Reviewer: do these automatically add the corresponding github label to the issue?

Author: yes!

---

> <!-- Lines like this are comments and will be invisible -->

Reviewer 2: You need to add this line in every template.

> <!-- Lines like this are comments and will be invisible -->

Author: I don't understand why you deem this necessary. As the comments are invisible, users *don't* need to remove them: That was the idea - instead of having lots of text in the issue template that people need to delete (or it clutters the issue if they don't), just some helpfully guiding comments. Also, I don't think users should face all-caps when they're already kind enough to give us some feedback in issues. I think we should make it as pleasant as possible to leave us an issue - at least to me, the proposed change sounds like we don't have a particularly high opinion of our users - definitely not something I'd like to read if I am about to give a project feedback.

> <!-- Lines like this are comments and will be invisible -->

Reviewer 2: > Also, I don't think users should face all-caps when they're already kind enough to give us some feedback in issues.

More text - less likely people read it, that is proven by using other templates and in many other projects. Not everything in the template is within HTML comments, and that text should be deleted. For me, the capitalized text is not insulting, it just a way of bringing attention to important bits in a text where you cannot use colour or more advanced approaches.

Why capitalized message can be important:

a. some people may not know what `<!-- -->` means, and we need to explain it.
b. some text requires removing and we need to shout about it.
c. all points above somehow should be emphasized in the text, so that a user reads it first.

> <!-- Lines like this are comments and will be invisible -->

Reviewer: I agree with ti here. Even if the intention's positive, it doesn't come across that way

> <!-- Lines like this are comments and will be invisible -->

Reviewer 2: Capitalized text is normal and used broadly. See another example: https://raw.githubusercontent.com/stevemao/github-issue-templates/master/bugs-only/ISSUE_TEMPLATE.md.

> <!-- Lines like this are comments and will be invisible -->

Author: At a): that's what this line already says. At b): what lines do you think need removing? The point of putting notes in comments was that they *don't* need removing... At c): I would prefer the risk of someone not noticing it (and perhaps leaving a sub-optimal issue report) over the risk of putting someone off from reporting an issue in the first place.
Someone else doing it that way doesn't necessarily make it right, or the best, way.

> <!-- Lines like this are comments and will be invisible -->

Author: ```suggestion
<!-- Lines like this, enclosed in <!-- -->, are comments and will be invisible -->
```

> <!-- Lines like this are comments and will be invisible -->

Reviewer: @awav I guess there's a subtlety to it, but as I say, what matters is how it reads not how we want it to read. There are cases where caps is OK, and I think the example you gave is OK, but I think you have to be careful what words you emphasise (by caps or punctuation) because of how it changes the tone

> # Bug / performance issue / build issue

Reviewer: At what point performance matters go from being bugs to being feature requests?

Author: Don't know - we've also not had many in the past. Something we can play by ear for now I think

> # Bug / performance issue / build issue

Reviewer: is the person writing the ticket meant to pick one of these and delete the others?

Author: if they want to - sure. if they don't - fine with me as well.

<!-- A clear and concise description of what the bug is. -->

## To reproduce

**Minimal, reproducible example**

> <!-- We need to be able to reproduce the bug by simply copy&pasting your code -->

Reviewer: ```suggestion
<!-- We need to be able to reproduce the bug by simply copy and pasting your code -->
```

```python
# This is the place for your code that reproduces the bug.

> # Please make sure it does not depend on external libraries, specific datasets, etc., and the smaller, the better :)

Reviewer: perhaps use the phrase 'minimal complete verifiable example'?

> # Please make sure it does not depend on external libraries, specific datasets, etc., and the smaller, the better :)

Reviewer: what about external libraries that GPflow uses? or do you not consider those external?

Author: The intent was to get people's MREs not to require *additional* dependencies beyond gpflow (and gpflow's own dependencies)

Reviewer: in that case, how about, instead of

> Please make sure it does not depend on external libraries

we have

>  Please make sure it does not depend on libraries that aren't also dependencies of GPflow

Author: ```suggestion
# Please make sure it does not depend on external libraries (beyond GPflow's own requirements), specific datasets, etc., and the smaller, the better :)
```

Reviewer: sounds good

> # For more information on good bug reports, check https://stackoverflow.com/help/minimal-reproducible-example

Reviewer: ```suggestion
# For help on how to write a good bug report, see https://stackoverflow.com/help/minimal-reproducible-example
```

```

> **Stack trace/error message**

Reviewer: ```suggestion
**Stack trace, or error message**
```
slashes don't read well

```

> // Paste the full stack trace/error message here

Reviewer: why `//` ?

Author: it's a comment - but then it's not python code? dunno. happy to change it to something else if you've got a better idea...

```

## Expected behavior

<!-- A clear and concise description of what you expected to happen. -->

## System information:

* GPflow version: <!-- run `import gpflow; print(gpflow.__version__)` inside python -->

> * GPflow installed from: <!-- e.g. with "pip install gpflow" or from a GitHub clone on the develop branch; if working from a git clone, please include the git commit (SHA hash): run `git rev-parse HEAD` on the command line -->

Reviewer: does the user need to specify the hash if they're using a tag?

Author: I guess tag would be fine, but tags are easily confused with branches, which keep moving

* TensorFlow version: <!-- run `import tensorflow; print(tensorflow.__version__)` inside python -->

> * Python version

Reviewer: ```suggestion
* Python version  <!-- run `python3 -V` -->
```

> * Python version  <!-- run `python3 -V` -->

Author: ```suggestion
* Python version  <!-- run `python3 -V` on the command line, or `import sys; print(sys.version)` inside python -->
```

Reviewer: :+1:
